src/transformers/utils/chat_template_utils.py

```diff
@@ -97,7 +98,7 @@ def _parse_type_hint(hint: str) -> Dict:
                 "Couldn't parse this type hint, likely due to a custom class or object: ", hint
             )
 
-    elif origin is Union:
+    elif origin in {Union, getattr(types, "UnionType", None)}:
```

Comment:
```suggestion
    elif origin is Union or (hasattr(types, "UnionType") and origin is types.UnionType):
```
This looks good now! I made one change in the suggestion here. The reason is subtle and I'm not sure it matters, but the `in` operator compares by equality `==`, not identity `is`. In other words, `x in (a, b)` is equivalent to `x == a or x == b`, not `x is a or x is b`, which is what we want here.

One final request: Can you add a test in `tests/utils/test_chat_template_utils.py` with a function that uses type hints with `| None`, so we can verify that this works and doesn't regress later?

Reply:
Sure, I'll do that tomorrow.
But your CI tests are running with multiple Python versions, including 3.9 no?
Won't the test break in Python 3.9 as the type hint format isn't recognized as valid Python?

Reply:
Ah, you're totally right! I forgot `X | Y` types were only added in 3.10 as well.

In that case, we probably can't test this until 3.9 is retired - it's annoying, but there's not much we can do.

Reply:
Alright, applied your suggestion and didn't add the test then!
I'll use this feature in my [company's project](https://github.com/Libertai/libertai-agents), I'll try to monitor it and submit another PR in case there's a regression 👍